src/NHibernate/AdoNet/AbstractBatcher.cs

```diff
@@ -242,37 +238,31 @@ public virtual DbDataReader ExecuteReader(DbCommand cmd)
 			CheckReaders();
 			LogCommand(cmd);
 			Prepare(cmd);
-			Stopwatch duration = null;
-			if (Log.IsDebugEnabled())
-				duration = Stopwatch.StartNew();
-			DbDataReader reader = null;
+
+			var duration = Log.IsDebugEnabled() ? Stopwatch.StartNew() : null;
+
+			var reader = DoExecuteReader(cmd);
+
+			_readersToClose.Add(reader);
+			LogOpenReader(duration , reader);
+			return reader;
+		}
+
+		private DbDataReader DoExecuteReader(DbCommand cmd)
+		{
 			try
 			{
-				reader = cmd.ExecuteReader();
+				var reader = cmd.ExecuteReader();
+				return _factory.ConnectionProvider.Driver.SupportsMultipleOpenReaders
+					? reader
+					: NHybridDataReader.Create(reader);
```

Comment:
Now `NHybridDataReader.Create` creation is also measured for `ExecuteReader took`. But why shouldn't it? It also allows to drop ugly `GetReaderStopwatch` with `NHybridDataReader` typing logic.